### ISSUE TITLE
Wifi chip from Smart Plug reports as Rtl8710c How to flash or dump it?

### ISSUE THREAD

**Issue by raufaser** [libretiny-eu/libretiny#266]
I desoldered the Wifi module from a Gosund EP2.
I managed to hook it up and It talks to me via UART.
But i have no idea how to put it into download mode and what kind of firmware i need.

This is the comm PCB: Front with W701 chip
![DSCF1030](https://github.com/libretiny-eu/libretiny/assets/72879802/13db31a9-3588-4d77-82ca-76abf41b0290)

And back:
![DSCF1039](https://github.com/libretiny-eu/libretiny/assets/72879802/4db5558d-6cce-402a-a191-460faa649279)

I hooked it up like this:
![Communication](https://github.com/libretiny-eu/libretiny/assets/72879802/56ba8441-0532-4b39-bd0f-670cc53a7d74)

And got this:
```
== Rtl8710c IoT Platform ==
Chip VID: 5, Ver: 3
ROM Version: v3.0

== Boot Loader ==
Dec  5 2019:14:02:18

fwx SELE[fffffffc]
fw SELE Bitidx 2, fw1 valid 1, sn 0, fw2 valid 1, sn 100
fw1 USE, return sn 0

Boot Loader <==


== RAM Start ==

Build @ 15:38:41, Jan 29 2023

interface 0 is initialized

interface 1 is initialized
...
...
See the txt for the whole capture
```

[capture.txt](https://github.com/libretiny-eu/libretiny/files/14621539/capture.txt)

Can someone help?

**Comment by raufaser** [comment]
I traced the chip pins to the outer pins. This is what i got:
![Pinout](https://github.com/libretiny-eu/libretiny/assets/72879802/e2ac3ff8-08aa-4696-bab2-70ad75582196)
I guessed a little and it seems to be the same as this one:
![363f707b-6210-4301-b2cf-f870d42cb821](https://github.com/libretiny-eu/libretiny/assets/72879802/eb34d798-2c9b-498a-947e-3d3fe1a1ddb8)

So i pulled A0 and CEN to 3,3V. Now i get this message:

```
== Rtl8710c IoT Platform ==
Chip VID: 5, Ver: 3
ROM Version: v3.0
Test Mode: boot_cfg1=0x20
Download Image over UART2[tx=16,rx=15] baud=115200
```

I am already communication via pins 16 and 15. Do you have any idea how to dump the firmware? Or what firmware i could flash?

**Comment by kuba2k2** [comment]
Try ltchiptool, it can dump and flash firmware of the chip. A ROM dump could be useful if you can share it.

As for firmware, LibreTiny has preliminary support for RTL8720C. Things like OTA are not implemented yet.

**Comment by raufaser** [comment]
I am a little bit confused, because my device reports as Rtl87**1**0c IoT Platform == ... is it different from RTL87**2**0C ?

No success yet with dumping.

**Comment by kuba2k2** [comment]
No, it's fundamentally the same thing. 8710 is the generic family name, also 8710 is a version of 8720 without Bluetooth. What matters is the "C" after the number.

**Comment by le-mon** [comment]
@makko81 can you tell me which Pins are RX2/TX2?

i already managed to get to:
```
== Rtl8710c IoT Platform ==
Chip VID: 5, Ver: 3
ROM Version: v3.0
Test Mode: boot_cfg1=0x20
Download Image over UART2[tx=16,rx=15] baud=115200
```

but ltchiptool doesn't do anything: 
```
❯ ltchiptool -vvv -s flash info -d /dev/ttyUSB0 ambz2
D: Found Local data snapshot (v1.13.0) - /home/user/.local/lib/python3.14/site-packages/ltchiptool
I: Connecting to 'Realtek AmebaZ2' on /dev/ttyUSB0 @ 115200
D: Linking: stage 0
S: <- TX: 'Rtk8710C'
V: push_timeout(0.1)
V: pop_timeout()
S: <- TX: 'ping'
S: <- TX: 'ping'
D: Linking: stage 1
S: <- TX: 'Rtk8710C'
V: push_timeout(0.1)
V: pop_timeout()
S: <- TX: 'ping'
D: Linking: stage 2
S: <- TX: 'Rtk8710C'
V: push_timeout(0.1)
V: pop_timeout()
S: <- TX: 'ping'
S: <- TX: 'ping'
D: Linking: stage 3
S: <- TX: 'Rtk8710C'
V: push_timeout(0.1)
V: pop_timeout()
S: <- TX: 'ping'
S: <- TX: 'ping'
S: <- TX: 'ping'
```

**Comment by kuba2k2** [comment]
As you can see from the logs, TX2 is PA16 and RX2 is PA15.

Your adapter's TX -> RX2 line might be not connected properly if you can see logs but can't send any commands.

**Comment by le-mon** [comment]
> TX -> RX2 line might be not connected properly

you were right, i changed TX & RX and at least it went one step further: 
```
❯ ltchiptool -vvv -s flash info -d /dev/ttyUSB0 ambz2
D: Found Local data snapshot (v1.13.0) - /home/user/.local/lib/python3.14/site-packages/ltchiptool
I: Connecting to 'Realtek AmebaZ2' on /dev/ttyUSB0 @ 115200
D: Linking: stage 0
S: <- TX: 'Rtk8710C'
V: push_timeout(0.1)
S: -> RX: 'Rtk8710C'
V: pop_timeout()
S: <- TX: 'ping'
V: push_timeout(0.048)
S: -> RX: 'ping'
S: <- TX: 'DW 40000038 4'
S: -> RX: 'DW 40000038 4'
E: ValueError: invalid literal for int() with base 16: 'DW'
E: |-- File "/home/user/.local/lib/python3.14/site-packages/ltchiptool/soc/ambz2/util/ambz2tool.py", line 188, in dump_words
```

I see you @kuba2k2 already know the problem as in https://github.com/libretiny-eu/ltchiptool/issues/63 but i am confused: I AM in Download mode, or did i something wrong? Are there some different modes?

**Comment by kuba2k2** [comment]
Yes, it seems like you are in download mode - at least that's what your first log snippet says: `Download Image over UART2[tx=16,rx=15] baud=115200`.

However, swapping RX and TX is not the way to go. Whatever you did right now most likely just makes the UART adapter echo back the commands that are being sent.

When you enter download mode, try opening a serial terminal (minicom, picocom, or similar) with baud rate 115200. Try typing something - you should NOT see anything being displayed. When you enter `ping` and press Enter, it should respond with `ping` as well. Same goes for `Rtk8710C`, but characters should NOT be visible as you type them (in real time).

**Comment by le-mon** [comment]
characters are shown as i type :(

btw: swapping RX/TX doesn't do any effect. strange...

**Comment by kuba2k2** [comment]
That just means your RX/TX wires are shorted together. There's no other explanation, besides a broken UART adapter. Try a different one, or rewire everything from scratch.

**Comment by le-mon** [comment]
Thats it!
> RX/TX wires are shorted together

I didn't know it is possible to have bad RX/TX wiring and still getting clear logs. 
One pad went missing so i had to use a needle to connect A15 directly next to the chip. Of course the tip of the needle is wide enough to also touch A16. 
After proper positioning I was able to read chip info. 

<img width="2000" height="1500" alt="Image" src="https://github.com/user-attachments/assets/9f896845-2b51-48dd-9724-89a33ea91736" />
